app/services/service_listeners/attachment_updater.rb

```diff
@@ -0,0 +1,67 @@
+module ServiceListeners
+  class AttachmentUpdater
+    def initialize(attachable)
+      @attachable = attachable
+    end
+
+    def update!
+      # if an attachment references an attachable, the attachable may
+      # not reference the attachment: for example, deleted attachments
+      # reference their attachable, but the attachable does not
+      # reference them.
+      @attachable.attachables.each do |attachable|
+        Attachment.where(attachable_id: attachable.id).find_each do |attachment|
```

Comment:
I reckon you should be able to do `Attachment.where(attachable: attachable.attachables)` here and ActiveRecord will turn that into a SQL `attachable_id IN (...)` query.